hive/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveTableOperations.java

```diff
@@ -0,0 +1,226 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive.legacy;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataOperations;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.hadoop.HadoopFileIO;
+import org.apache.iceberg.hive.HiveClientPool;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class LegacyHiveTableOperations extends BaseMetastoreTableOperations {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LegacyHiveTableOperations.class);
+
+  private final HiveClientPool metaClients;
+  private final String database;
+  private final String tableName;
+  private final Configuration conf;
+
+  private FileIO fileIO;
+
+  protected LegacyHiveTableOperations(Configuration conf, HiveClientPool metaClients, String database, String table) {
+    this.conf = conf;
+    this.metaClients = metaClients;
+    this.database = database;
+    this.tableName = table;
+  }
+
+  @Override
+  public FileIO io() {
+    if (fileIO == null) {
+      fileIO = new HadoopFileIO(conf);
+    }
+
+    return fileIO;
+  }
+
+  @Override
+  protected void doRefresh() {
+    try {
+      org.apache.hadoop.hive.metastore.api.Table hiveTable =
+          metaClients.run(client -> client.getTable(database, tableName));
+
+      Schema schema = LegacyHiveTableUtils.getSchema(hiveTable);
+      PartitionSpec spec = LegacyHiveTableUtils.getPartitionSpec(hiveTable, schema);
+
+      TableMetadata metadata = TableMetadata.newTableMetadata(schema, spec, hiveTable.getSd().getLocation(),
+          LegacyHiveTableUtils.getTableProperties(hiveTable));
+      metadata = metadata.replaceCurrentSnapshot(getDummySnapshot());
+      setCurrentMetadata(metadata);
+    } catch (TException e) {
+      String errMsg = String.format("Failed to get table info from metastore %s.%s", database, tableName);
+      throw new RuntimeException(errMsg, e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted during refresh", e);
+    }
+    setShouldRefresh(false);
```

Comment:
Is the expectation that once the metadata is read once, subsequent reads in the application will still see the same metadata?
In normal legacy Hive code path, metadata can dynamically evolve through subsequent reads. Do you think the behavior here is backward compatible?

Reply:
When you use the term metadata, I think there are two things here, table level metadata (which includes table schema, properties, etc) and files returned by the table scans. Once a table object is created, the metadata remains same. However, if new files/partitions are added, they will be included in the scan planning. This is compatible with existing Hive behaviour.